### PR TITLE
Ray Tune Integration Updates

### LINKED ISSUE #11249: TypeError: can't pickle _thread.RLock objects hyperparameter_search raytune
## Environment info

- `transformers` version: v4.5.1
- Platform: Linux
- Python version: 3.7.8
- PyTorch version (GPU?): yes
- Tensorflow version (GPU?):
- Using GPU in script?: 
- Using distributed or parallel set-up in script?:

## Information

Model I am using (Bert, XLNet ...): bert-base-uncased 

The problem arises when using:
* [ x] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ x] my own task or dataset: (give details below)

## To reproduce

1. Run hyperparameter tuning with raytune
2.
3.

```
2021-04-14 15:44:01,389 INFO services.py:1264 -- View the Ray dashboard at http://127.0.0.1:8265
Traceback (most recent call last):
  File "pipeline_training.py", line 311, in <module>
    keep_checkpoints_num=0
  File "/opt/conda/lib/python3.7/site-packages/transformers/trainer.py", line 1459, in hyperparameter_search
    best_run = run_hp_search(self, n_trials, direction, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/transformers/integrations.py", line 235, in run_hp_search_ray
    **kwargs,
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/tune.py", line 297, in run
    _ray_auto_init()
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/tune.py", line 664, in _ray_auto_init
    ray.init()
  File "/opt/conda/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 62, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 785, in init
    hook()
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/registry.py", line 171, in flush
    self.references[k] = ray.put(v)
  File "/opt/conda/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 62, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 1481, in put
    object_ref = worker.put_object(value)
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 266, in put_object
    serialized_value = self.get_serialization_context().serialize(value)
  File "/opt/conda/lib/python3.7/site-packages/ray/serialization.py", line 324, in serialize
    return self._serialize_to_msgpack(value)
  File "/opt/conda/lib/python3.7/site-packages/ray/serialization.py", line 304, in _serialize_to_msgpack
    self._serialize_to_pickle5(metadata, python_objects)
  File "/opt/conda/lib/python3.7/site-packages/ray/serialization.py", line 264, in _serialize_to_pickle5
    raise e
  File "/opt/conda/lib/python3.7/site-packages/ray/serialization.py", line 261, in _serialize_to_pickle5
    value, protocol=5, buffer_callback=writer.buffer_callback)
  File "/opt/conda/lib/python3.7/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/opt/conda/lib/python3.7/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 580, in dump
    return Pickler.dump(self, obj)
TypeError: can't pickle _thread.RLock objects
```

The code chunk to start the `hyperparameter_search`: 

```python
def my_hp_space(trial):
    from ray import tune
    return {
        "learning_rate": tune.uniform(1e-5, 5e-5),
        "num_train_epochs": tune.choice(range(1, 6)),
        "per_device_train_batch_size": tune.choice([2,4]),
        "weight_decay": tune.uniform(0.0, 0.3),
        "adam_epsilon": tune.loguniform(1e-10, 1e-6),
        "per_device_eval_batch_size": 32
    }
    
best_run = trainer.hyperparameter_search(
    backend="ray",
    n_trials=15,
    hp_space=my_hp_space,
    stop=None,    
    checkpoint_score_attr="training_iteration",
    keep_checkpoints_num=0
    compute_objective=lambda x: my_objective(x, metric='eval_' + used_metric)
)
```

## Expected behavior

Expect that it will not throw an error. Note that this script does work on `4.2.0`.

### PR DESCRIPTION
**Pull request by amogkam**
# What does this PR do?
- Automatically disables memory tracker if enabled since the memory tracker is not serializable
- Fixes the Ray Tune integration test
- Adds a new test for Ray Client API
- Adds integration tests back to the scheduled Github Actions pipeline

Closes #11249, https://github.com/huggingface/transformers/issues/12177 

@LysandreJik @richardliaw

### CODE REVIEW

File: tests/test_trainer.py [hunk 1]
```diff
@@ -1323,3 +1329,14 @@ def hp_name(params):
             trainer.hyperparameter_search(
                 direction="minimize", hp_space=hp_space, hp_name=hp_name, backend="ray", n_trials=4
             )
+
+    def test_hyperparameter_search(self):
+        self.ray_hyperparameter_search()
```
Comment (richardliaw) [inline tests/test_trainer.py:1334]: should this return?

**Review by richardliaw — COMMENTED** [review]
looks good to me.

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -163,11 +163,21 @@ def _objective(trial, local_trainer, checkpoint_dir=None):
...
 
+    if not trainer._memory_tracker.skip_memory_metrics:
+        from .trainer_utils import TrainerMemoryTracker
+
+        logger.warning(
+            "Memory tracking for your Trainer is currently "
+            "enabled. Automatically disabling the memory tracker "
+            "since the memory tracker is not serializable."
+        )
+        trainer._memory_tracker = TrainerMemoryTracker(skip_memory_metrics=True)
```
Comment (richardliaw) [inline src/transformers/integrations.py:174]: @LysandreJik maybe we can have this as a more generic API? nice to have, non blocking

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

File: src/transformers/integrations.py [hunk 3]
```diff
@@ -194,7 +204,8 @@ def _objective(trial, local_trainer, checkpoint_dir=None):
         trainer.use_tune_checkpoints = True
         if kwargs["keep_checkpoints_num"] > 1:
             logger.warning(
-                f"Currently keeping {kwargs['keep_checkpoint_num']} checkpoints for each trial. "
+                f"Currently keeping {kwargs['keep_checkpoints_num']} "
+                f"checkpoints for each trial. "
```
[the remark this answers is not in the document]
Comment (amogkam) [reply src/transformers/integrations.py]: Thanks @sgugger, made the change!

**Comment by richardliaw** [comment]
thanks @sgugger for the fast review! anything blocking to get this merged :) ?

**Comment by sgugger** [comment]
This is good for me. You had standing questions for Lysandre so not sure it was ready to be merged, but I will do so if you tell me everything is okay on your side :-)

**Comment by amogkam** [comment]
@sgugger yep this is ready to merge!

**Comment by sgugger** [comment]
Thanks again!

**Pull request merged**